### PR TITLE
New macros for CBMC contracts

### LINKED ISSUE #169: Make CBMC proofs more readable
Currently, the CBMC annotations in the code are difficult to read and confusing to readers. For example:

```c
#define poly_decompress KYBER_NAMESPACE(poly_decompress)
void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
/* *INDENT-OFF* */
__CPROVER_requires(a != NULL)
__CPROVER_requires(__CPROVER_is_fresh(a, KYBER_POLYCOMPRESSEDBYTES))
__CPROVER_requires(r != NULL)
__CPROVER_requires(__CPROVER_is_fresh(r, sizeof(poly)))
__CPROVER_assigns(__CPROVER_object_whole(r))
__CPROVER_ensures(__CPROVER_forall { unsigned k; (k < KYBER_N) ==> ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q ) });
/* *INDENT-ON* */

void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
{
    #if (KYBER_POLYCOMPRESSEDBYTES == 128)
    for (int i = 0; i < KYBER_N / 2; i++)
        __CPROVER_assigns(i, __CPROVER_object_whole(r))
        __CPROVER_loop_invariant(i >= 0)
        __CPROVER_loop_invariant(i <= KYBER_N / 2)
        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
                                 ( r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
                                   r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q ) })
        __CPROVER_decreases(KYBER_N / 2 - i)
    {
        // REF-CHANGE: Hoist scalar decompression into separate function
        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[i] >> 0) & 0xF);
        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[i] >> 4) & 0xF);
    }



    #elif (KYBER_POLYCOMPRESSEDBYTES == 160)
    const int num_blocks = KYBER_N / 8;
    for (int i = 0; i < num_blocks; i++)
        __CPROVER_assigns(i, __CPROVER_object_whole(r))
        __CPROVER_loop_invariant(i >= 0)
        __CPROVER_loop_invariant(i <= num_blocks)
        __CPROVER_loop_invariant(num_blocks == 32)
        __CPROVER_loop_invariant(__CPROVER_forall { int k; k >= 0 && k <= (8 * i - 1) ==>
                                 ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q )})
        __CPROVER_decreases(num_blocks - i)
...
```

- [ ] Introduce some abbreviations for the various `__CPROVER` macros to make them less daunting to read.

In addition to plain abbreviations for `__CPROVER_xxx`, one could also introduce macros for repeating bounds conditions like `_CPROVER_forall { int k; k >= 0 && k <= (8 * i - 1) ==> ( r->coeffs[k] >= 0 && r->coeffs[k] < KYBER_Q )}` above.

### PR DESCRIPTION
**Pull request by rod-chapman**
This PR provides a draft proposal for new macros to make the CBMC contracts less verbose and easier to read.

Fixes #169 

New macros are in mlkem/cbmc.h

I went for UPPERCASE for the new macros, since this appears to be a common convention in C, and also minimizes the change of creating name-clashes in the future.

I have applied these in mlkem/poly.[hc] to see how they look. All for discussion.

Tests and proofs all pass on this branch locally.

### CODE REVIEW

File: mlkem/cbmc.h [hunk 1]
```diff
@@ -1,10 +1,86 @@
...
+///////////////////////////////////////////////////
+
+// Boolean-value predidate that asserts that "all values of array_var are in range value_lb .. value_ub (inclusive)"
+//
+// Example:
+//  ARRAY_IN_TYPE(unsigned, k, 0, KYBER_N-1, a->coeffs, -QM1, QM1)
+// expands to
+//  __CPROVER_forall { unsigned k; (0 <= k && k <= KYBER_N-1) ==> ( (-QM1 <= a->coeffs[k]) && (a->coeffs[k] <= QM1)) }
+
+#define ARRAY_IN_TYPE(indextype, qvar, qvar_lb, qvar_ub, array_var, value_lb, value_ub) __CPROVER_forall { indextype qvar; (qvar_lb <= qvar && qvar <= qvar_ub) ==> ( (value_lb <= array_var[qvar]) && (array_var[qvar] <= value_ub)) }
```
Comment (hanno-becker) [inline mlkem/cbmc.h:84]: Nit: The name is not very intuitive. Change to `ARRAY_IN_BOUNDS` instead?

File: mlkem/params.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 #define KYBER_N 256
 #define KYBER_Q 3329
+#define QM1     3328 // "Q minus 1"
```
Comment (hanno-becker) [inline mlkem/params.h:25]: I'm not sure this is needed?

File: mlkem/cbmc.h [hunk 3]
```diff
@@ -1,10 +1,86 @@
...
+
+// History variables
+// https://diffblue.github.io/cbmc/contracts-history-variables.html
+#define OLD(...)             __CPROVER_old(__VA_ARGS__)
+#define LOOP_ENTRY(...)      __CPROVER_loop_entry(__VA_ARGS__)
+
+// Quantifiers
+// Note that the range on qvar is _inclusive_ between qvar_lb .. qvar_ub
+// https://diffblue.github.io/cbmc/contracts-quantifiers.html
+#define FORALL(type, qvar, qvar_lb, qvar_ub, predicate) __CPROVER_forall { type qvar; (qvar_lb <= qvar && qvar <= qvar_ub) ==> ( predicate ) }
```
Comment (hanno-becker) [inline mlkem/cbmc.h:70]: Here and in many other places: Always use brackets around expression arguments, e.g. 
```suggestion
#define FORALL(type, qvar, qvar_lb, qvar_ub, predicate) __CPROVER_forall { type qvar; (qvar_lb) <= (qvar) && (qvar) <= (qvar_ub)) ==> ( predicate ) }
```

File: mlkem/poly.c [hunk 4]
```diff
@@ -108,13 +109,12 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
...
-        __CPROVER_loop_invariant(i >= 0)
-        __CPROVER_loop_invariant(i <= KYBER_N / 2)
-        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
-                                 ( r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
-                                   r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q ) })
-        __CPROVER_decreases(KYBER_N / 2 - i)
+        ASSIGNS(i, OBJECT_WHOLE(r))
+        INVARIANT(i >= 0)
+        INVARIANT(i <= KYBER_N / 2)
+        INVARIANT(FORALL(int, k, 0, (i-1), (r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
```
Comment (hanno-becker) [inline mlkem/poly.c:115]: Can this not be formulated using `ARRAY_IN_TYPE`?

Comment (rod-chapman) [reply mlkem/poly.c:115]: I don't think so... the quantified variable is "k" but we want "2 * k" and so on inside the predicate on the right, so it wouldn't expand correctly using the macro.

File: mlkem/cbmc.h (hunk 1 above)
Comment (rod-chapman) [reply mlkem/cbmc.h]: Agreed. Done.

File: mlkem/poly.c (hunk 4 above)
Comment (hanno-becker) [reply mlkem/poly.c:115]: You'd need to double the upper bound index, yes, but wouldn't that work?

File: mlkem/cbmc.h (hunk 3 above)
Comment (rod-chapman) [reply mlkem/cbmc.h]: Done.

File: mlkem/poly.c (hunk 4 above)
Comment (rod-chapman) [reply mlkem/poly.c:115]: Don't think so... if the quantified variable is "k", then the macro produces r->coeffs[k] on the right, not r->coeffs[2 * k]

File: mlkem/params.h (hunk 2 above)
Comment (rod-chapman) [reply mlkem/params.h]: Done

File: mlkem/poly.c (hunk 4 above)
Comment (hanno-becker) [reply mlkem/poly.c:115]: But you have two clauses, one for `2k` and one for `2k+1` -- so overall you bound _all_ array elements..

File: mlkem/poly.c [hunk 5]
```diff
@@ -108,13 +109,12 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
...
+                                            r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q)))
```
Comment (hanno-becker) [inline mlkem/poly.c:116]:
```suggestion
        INVARIANT(FORALL(int, k, 0, 2*(i-1), (r->coeffs[k]     >= 0 && r->coeffs[k]     < KYBER_Q)))
```

Comment (rod-chapman) [reply mlkem/poly.c:116]: I see ... let me try that...

Comment (rod-chapman) [reply mlkem/poly.c]: Yes... works. I also did it using ARRAY_IN_BOUNDS rather than FORALL.

File: mlkem/poly.c (hunk 4 above)
Comment (rod-chapman) [reply mlkem/poly.c]: OK... got it and it works.

File: mlkem/poly.c (hunk 5 above)
Comment (hanno-becker) [reply mlkem/poly.c]: Nice!

**Review by hanno-becker — APPROVED** [review]

**Pull request merged**